torch/_inductor/codegen/cpp_micro_gemm.py

```diff
@@ -373,6 +392,10 @@ class CppMicroGemmFP32Vec(CppMicroGemm):
     Vectorized va;
     at::vec::VectorizedN<{{compute_t}}, COLS> vb;
     at::vec::VectorizedN<{{compute_t}}, ROWS*COLS> vc;
+    {%- if input2_dtype == torch.int8 %}
+    // Load up to max_num_weight_registers registers of B at a time for int32 -> fp32 conversion.
+    constexpr int max_num_weight_registers = 64 / VLEN;
```

Comment:
With AVX2, there would be 8 registers while there are only 16 registers in total. Would 8 registers be too many for AVX2?

Reply:
Thanks for reviewing!

Currently, up to 3 registers are loaded with FP32 weights (after int8 -> int32 -> fp32 conversion) because `BLOCK_N` is 48 for AVX512, and 24 for AVX2. I mentioned in the comments that it's okay for now, but may have to change later. However, on second thought, I think i should change it now.

Before my last commit, up to 4 ZMM vectors corresponding to B were being loaded at once. Practically, up to 3 registers are loaded with FP32 weights (after int8 -> int32 -> fp32 conversion) because `BLOCK_N` is 48 for AVX512, and 24 for AVX2. I mentioned in the comments that it's okay for now, but may have to change later.

Even now, though, it may cause register spilling if `compute` would be unrolled -  one invocation of `compute` with the same `k` uses 1 register for `C`, one for `A`. If we use `3` for `B` (preloading all necessary `B` elements at once), if `compute` would be unrolled 4 times, that would mean 20 registers being needed, so in case of AVX2, one B vector would have to be spilled for each invocation! That seems to indicate that for AVX2, a maximum of 2 `B` vectors should be loaded together.

Reply:
@jgong5, please confirm if `compute` lambda is actually being unrolled. It uses `vb`, which is outside it, so is it possible that a compiler may not unroll it due to concerns of memory aliasing (not sure memory aliasing is an issue for compilers  if loads from different iterations are from same addresses, though, but since aliasing would definitely be an issue when `k` would change, I suspect the compiler wouldn't unroll the `compute` lambda) ?

Since `i` would vary from 0 to `ROWS * COLS`, for each value of `k`, for AVX512, may we load `vb` at the beginning of each k iteration?
Practically, `vb` is just 3 YMM/ZMM vectors, and would be constant across `ROWS * COLS` iterations of `compute` lambda (and thus can be shared among unrollings of the `compute` lambda as well).

Thanks!

Reply:
Hi @jgong5 @leslie-fang-intel, I moved loading of `vb` vectors outside the `compute` lambda. It's being done for each value of `k` now.

BTW, `ForcedUnroll` only seems to invoke `compute` some specific number of times. Would this explicit unrolling result in the same behavior as a `#pragma unroll` of a loop with regards to Instruction-level parallelism?

Thanks!

Reply:
Could you please avoid doing this change in this PR? If you think moving the load of vb out is better, please do that in the next PR with perf numbers and code sequences proving it doesn't hurt perf for bf16/fp16/woq.

Reply:
Perf seems to be same, but as you advised, I could do that in a subsequent PR.

Reply:
Perf is better (e.g. for M=124, K=128, N=1024) when `vb` is loaded one ZMM/YMM vector at a time. Reverted to such an implementation. Thanks!

Reply:
Hi @jgong5, to clarify - the very first implementation in this PR loaded one FP32 weight vector at a time (from 16 or 8 bytes of int8 weights, depending upon AVX512/AVX2 ISA).

The reason why I now reverted to that approach instead of following through on [your review comments about loading 64 int8 elements](https://github.com/pytorch/pytorch/pull/131887/commits/306970b35b631f44466975b2e13a3904c8fa977e#r1694790290) is that the former approach works better, perhaps because it intersperses memory loads with compute better than the latter approach.

Reply:
Thanks for the explanation! The very first implementation looks good to me. It is simpler too.